### PR TITLE
Preserve data

### PR DESCRIPTION
**Pull request by ifeanyi305**
Hi reviewer, In this PR my coding partner and I made the following changes:

- On the exit from the app, all the data are saved (books, people, and rentals) into JSON files. One per type of data.
- On startup of the app, it loads all data from files. we handled the case when one or more of the files are missing.

### CODE REVIEW

File: serializer.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    end
+    serialized_data = JSON.dump(data)
+    File.write("#{@path}/#{file_name}.json", serialized_data)
+  end
+
+  def apply_rentals(books, people, rentals)
+    rentals.each do |rental|
+      book = books.find { |current_book| current_book.title == rental['book'] }
+      person = people.find { |current_person| current_person.id == rental['person'] }
+      Rental.new(rental['date'], book, person)
```
Comment (omar-muhamad) [inline serializer.rb:62]: - [ ] There is an error raised at the app start, Please fix that.

image

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @ifeanyi305 and @Kweeka1,

Nice job, you are almost there.



## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- You used the correct workflow.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: serializer.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  def load(file_name)
+    data = []
+    begin
+      deserialized_data = JSON.load_file("#{@path}/#{file_name}.json")
+      deserialized_data.each do |object|
+        case file_name
+        when 'books' then data << Book.new(object['title'], object['author'])
+        when 'students'
+          data << Student.new(object['age'], object['id'], object['name'],
+                              parent_permission: object['parent_permission'])
```
Comment (omar-muhamad) [inline serializer.rb]: - [ ] Same error is raised when I saved student data and try to re-run the App, Please fix that.

image

[Hint] The problem is fixed by rearranging the student arguments and switching places between id and name.

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @ifeanyi305 and @Kweeka1,

Nice job making the required changes. Unfortunately, you missed some of them but you are almost there.



## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- You used the correct workflow. ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by Ghiftee — APPROVED** [review]
## STATUS: APPROVED ♻️♻️
Hi @ifeanyi305 and @Kweeka1 , 

To highlight 👍🏽 
✔️ App runs smoothly
✔️ You added features for user experience
✔️ No linter errors

Your project is complete!  There is nothing else to say other than... it's time to merge it 🚀  
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**Pull request merged**